### PR TITLE
prov/efa: EFA send path/MR fixes

### PR DESCRIPTION
**Pull request by rwespetal**
There is a bug in the EFA provider that causes the memory descriptor to be skipped when specified by the app, causing an unneeded memory copy in the send path. Do some other cleanup while we're here. There is more to do but this at least fixes the issue for now.

Test description: Added some prints and ran `fi_rdm_tagged_pingpong`/`fi_rma_bw -o write` to verify the descriptor is being used for the various protocols. Did some testing with MPI (latency, bw, message rate, alltoall) as well and verified inline registration is still functional.

### CODE REVIEW

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 1]
```diff
@@ -84,14 +78,7 @@ ssize_t rxr_pkt_post_data(struct rxr_ep *rxr_ep,
...
 
-	/*
-	 * TODO: Check to see if underlying device can support CUDA
-	 * registrations and fallback to rxr_ep_send_data_pkt_entry() if it does
-	 * not. This should be done at init time with a CUDA reg-and-fail flag.
-	 * For now, always send CUDA buffers through
-	 * rxr_pkt_send_data_desc().
-	 */
-	if (efa_is_cache_available(efa_domain) || efa_ep_is_cuda_mr(tx_entry->desc[0]))
+	if (tx_entry->desc[0])
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:81]: Don't we still need to enter here during the cache path?

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr.h [hunk 2]
```diff
@@ -969,14 +969,6 @@ static inline struct rxr_domain *rxr_ep_domain(struct rxr_ep *ep)
 	return container_of(ep->util_ep.domain, struct rxr_domain, util_domain);
 }
 
-static inline uint8_t rxr_ep_mr_local(struct rxr_ep *ep)
-{
-	struct rxr_domain *domain = container_of(ep->util_ep.domain,
-						 struct rxr_domain,
-						 util_domain);
-	return domain->mr_local;
```
[the remark this answers is not in the document]
Comment (rwespetal) [reply prov/efa/src/rxr/rxr.h:977]: Updated.

File: prov/efa/src/rxr/rxr_pkt_cmd.c (hunk 1 above)
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:81]: No. `rxr_ep_tx_init_mr_desc` sets `tx_entry->desc` after doing the inline registration. The only cache check should be when we decide to do inline registration or not and no where else in the send path. I haven't been able to make those changes yet.

**Comment by rwespetal** [comment]
NCCL [single node](https://libfabric-ci.aws.a2z.com/job/nccl-single-node/1329/PROVIDER=efa,label=ubuntu_18.04/console) failure is suspicious:

```
16:24:28 WARN: Function: allocate_buff Line: 103: NET/OFI Cuda failure 'out of memory'
16:24:28 WARN: Function: main Line: 154: NET/OFI OFI NCCL failure: 1
```

File: prov/efa/src/rxr/rxr_pkt_cmd.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:81]: We already call ```rxr_ep_prepare_desc_send``` in  ```rxr_pkt_handle_long_rtm_sent```, which will set ```tx_entry->desc[0]``` for the cache path.

**Review by wzamazon — APPROVED** [review]

**Comment by rwespetal** [comment]
> NCCL [single node](https://libfabric-ci.aws.a2z.com/job/nccl-single-node/1329/PROVIDER=efa,label=ubuntu_18.04/console) failure is suspicious:
> 
> ```
> 16:24:28 WARN: Function: allocate_buff Line: 103: NET/OFI Cuda failure 'out of memory'
> 16:24:28 WARN: Function: main Line: 154: NET/OFI OFI NCCL failure: 1
> ```

I can't reproduce the failure. I tracked it down to this line:

```
ncclResult_t allocate_buff(void **buf, size_t size, int buffer_type)
{
 [ .. snip .. ]
        case NCCL_PTR_HOST:
                NCCL_OFI_TRACE(NCCL_NET, "Allocating host buffer");
                CUDACHECK(cudaHostAlloc((void **)buf, size, cudaHostAllocMapped));
```

I'm not sure why `cudaHostAlloc` failed (it is allocating pinned host pages mapped in the CUDA address space) or how this change could cause that failure.

**Pull request merged**